modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk.clj

```diff
@@ -299,3 +301,24 @@
 (defmethod driver/db-start-of-week :bigquery-cloud-sdk
   [_]
   :sunday)
+
+(defn- populate-project-id-from-credentials
+  [{:keys [details] :as database}]
+  ;; update the details blob to include the credentials' project-id as a separate entry
+  ;; this is basically an inferred/calculated key (not something the user will ever set directly), since it's simply
+  ;; part of the service account JSON, which is a field they provide
+  ;; it would be a lot of extra computation/invocation of the Google SDK to recalculate this on every query (since
+  ;; it will involve parsing this JSON repeatedly), and even using something like qp.store/cached functionality would
+  ;; still require recomputing it on every query execution
+  ;; since this will only ever change if/when the details change (i.e. the service account), just calculate it once
+  ;; when the DB is updated and store it back to the app DB
+  (println "populate-project-id-from-credentials")
+  (db/update! Database
+    (u/the-id database)
+    :details
+    (assoc details :project-id-from-credentials (-> (database-details->service-account-credential details)
+                                                    (.getProjectId)))))
```

Comment:
You don't need to wrap Java interop calls in `()` with the threading macros, you can just do

```suggestion
                                                    .getProjectId))))
```

Reply:
Maybe this is a dumb idea, but did you consider deferring the calculating and saving stuff until we actually need it (i.e., when making a BQ REST API request?).

The reason I'm suggesting that is `notify-database-updated` only gets called when when someone _updates_ a Database from the API endpoint. If you _create_ a new Database, we're not going to end up calculating this info

Reply:
This would also fix the drama with your test since you wouldn't have to use `notify-database-updated`

Reply:
It feels a bit weird to me to be calling a Toucan model update from `driver/can-connect?` (essentially), but yeah that is maybe less complicated overall.

Reply:
OK, this change has been made (but it also is invoked from `notify-database-updated` since in theory, the service account JSON _can_ change and that would require recalculating it again.